src/orquestra/sdk/_base/_traversal.py

```diff
@@ -72,43 +71,195 @@ def _gen_id_hash(*args):
+
+        return artifact
 
-    def traverse(self, root_futures: t.Sequence[GraphNode]):
+    def traverse(self, output_nodes: t.Sequence[DSLDataNode]):
         """
-        Traverse the workflow graph.
+        Traverse the DSL workflow graph and collect the IR models.
 
         We iterate over the futures returned from the workflow find the artifacts,
         constants, and secrets.
         """
-        artifact_counter = 0
         secret_counter = 0
         constant_counter = 0
-        for n in _iter_nodes(root_futures):
+
+        seen_futures: t.MutableSet[_dsl.ArtifactFuture] = set()
+        seen_invocations: t.MutableSet[_dsl.TaskInvocation] = set()
+
+        for n in _iter_nodes(output_nodes):
             if isinstance(n, _dsl.ArtifactFuture):
-                self._artifacts[_make_key(n)] = _make_artifact_node(artifact_counter, n)
-                artifact_counter += 1
-                # Map the invocation to the future.
-                # Note: Each unique future has one invocation, but each invocation
-                #       can have many Futures.
-                #       We're mapping `invocation: set(futures from invocation)`
-                self._invocations.setdefault(n.invocation, set()).add(n)
+                # The main point of handling artifact futures is to populate:
+                # - self._future_artifacts
+                # - self._invocation_unpacked_artifacts
+                # - self._invocation_non_unpacked_artifacts
+
+                # A single future can be used in multiple places in the workflow. We
+                # only want to handle it once.
+                if n in seen_futures:
+                    continue
+
+                out_meta = n.invocation.task.output_metadata
+
+                # There can be many futures requiring to run the same task invocation.
+                if n.invocation in seen_invocations:
+                    # We've already handled the invocation, but not this future. We're
+                    # supposed to have artifact nodes already generated for it. We just
+                    # need to point "self._future_artifacts" to it.
+
+                    if out_meta.is_subscriptable:
+                        assert n.invocation in self._invocation_unpacked_artifacts
+
+                    assert n.invocation in self._invocation_non_unpacked_artifacts
+
+                    self._point_future_to_artifact(n)
+                else:
+                    # We haven't seen the invocation and haven't seen the future. We'll
+                    # have to generate artifact nodes it before we can point
+                    # "self._future_artifacts" to it.
+
+                    if out_meta.is_subscriptable:
+                        # Generate artifact nodes for unpacking ("bar" and "baz" in the
+                        # example above).
+                        unpacked_artifacts = []
+                        for output_i in range(out_meta.n_outputs):
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Do we now store all unpacked artifacts in the IR even if they aren't used?